pymongo/mongo_client.py

```diff
@@ -2304,6 +2204,166 @@ def __exit__(
+        retryable: Optional[bool] = None,
+    ):
+        self._last_error: Optional[Exception] = None
+        self._retrying = False
+        self._multiple_retries = _csot.get_timeout() is not None
+        self._client = mongo_client
+        self._retry_operation = (
+            mongo_client.options.retry_reads if is_read else mongo_client.options.retry_writes
+        )
+
+        self._func = func
+        self._bulk = bulk
+        self._session = session
+        self._is_read = is_read
+        self._retryable = retryable and self._retry_operation and not self._in_transaction()
+        self._read_pref = read_pref
+        self._server_selector = read_pref if is_read else writable_server_selector
+        self._address = address
+        self._server: Server = None
+
+    def run(self) -> T:
+        """Runs the supplied func() and attempts a retry
+
+        :raises self._last_error: Exception raised
+        :return: Result of the func() call
+        :rtype: T
+        """
+        # Increment the transaction id up front to ensure any retry attempt
+        # will use the proper txnNumber, even if server or socket selection
+        # fails before the command can be sent.
+        if not self._in_transaction() and self._retryable and not self._is_read:
+            self._session._start_retryable_write()
+            if self._bulk:
+                self._bulk.started_retryable_write = True
+
+        while True:
+            self._check_last_error(check_csot=True)
+            try:
+                self._server = self._client._select_server(
+                    self._server_selector, self._session, address=self._address
+                )
+                return self._read() if self._is_read else self._write()
+            except ServerSelectionTimeoutError:
+                # The application may think the write was never attempted
+                # if we raise ServerSelectionTimeoutError on the retry
+                # attempt. Raise the original exception instead.
+                self._check_last_error()
+                # A ServerSelectionTimeoutError error indicates that there may
+                # be a persistent outage. Attempting to retry in this case will
+                # most likely be a waste of time.
+                raise
+            except PyMongoError as exc:
+                # Execute specialized catch on read
+                if self._is_read:
+                    if isinstance(exc, (ConnectionFailure, OperationFailure)):
+                        exc_code = getattr(exc, "code", None)
+                        if self._is_not_retry_eligible() or (
+                            exc_code and exc_code not in helpers._RETRYABLE_ERROR_CODES
+                        ):
+                            raise
+                        self._retrying = True
+                        self._last_error = exc
+                    else:
+                        raise
+
+                # Assumed write operation henceforth
```

Comment:
It looks like we can get to this point if we get to the `self._retrying = True` case for a read above.

Reply:
Ah, you're right. Looks like I'll just add an explicit if conditional `if not self._is_write` above line 2285 to avoid this.